### ISSUE TITLE
Linking error on Linux

### ISSUE THREAD

**Issue by BurnySc2** [PyO3/pyo3#740]
I am developing from Linux Mint 19 using `cargo 1.40.0-nightly (5da4b4d47 2019-10-28)`
 and running into the following issue:

I copied the simple example code from https://pyo3.rs/v0.8.5/class.html#constructor
to an example repository:

https://github.com/BurnySc2/PyO3-error/blob/f5d68bd119b509a9aad2ad0a091d7abe08330d41/src/lib.rs#L13-L16

However, as soon as I enable the `new()` function with `#[new]` I get the following linking error on linux:

https://github.com/BurnySc2/PyO3-error/commit/f5d68bd119b509a9aad2ad0a091d7abe08330d41/checks?check_suite_id=408095792#step:9:67

On windows, it compiles without a problem.

I also tested upgrading from 0.8.5 to the 0.9.0-alpha.1 version while changing the syntax properly

https://github.com/BurnySc2/PyO3-error/blob/1c559d55332f7e6bf074f7cd771b24eb034db38a/src/lib.rs#L13-L16

The error is the same:

https://github.com/BurnySc2/PyO3-error/runs/397018311#step:9:126

The error can be reproduced using the following github actions workflow:

https://github.com/BurnySc2/PyO3-error/blob/master/.github/workflows/rust_build.yml

In summary:
`cargo test` and `cargo bench` do not work on linux, but `cargo build --release` does.
All three of the above commands work on windows.

**Comment by danielvschoor** [comment]
Just like to mention that I'm experiencing this exact issue as well on WSL Ubuntu, whereas I can successfully compile on Windows

**Comment by davidhewitt** [comment]
This is a duplicate of #341 - see the note on testing [here](https://github.com/PyO3/pyo3/blob/master/guide/src/advanced.md)

cc @kngwyu to please close (I don't have permissions!)

**Comment by BurnySc2** [comment]
Thanks for pointing me in the right direction!
